executor/common_linux.h

```diff
@@ -1355,6 +1355,190 @@ static long syz_emit_ethernet(volatile long a0, volatile long a1, volatile long
+
+	return 0;
+}
+
+#endif
+
+#if SYZ_EXECUTOR || __NR_syz_io_uring_submit
+
+static long syz_io_uring_submit(volatile long a0, volatile long a1, volatile long a2, volatile long a3)
+{
+	// syzlang: syz_io_uring_submit(sq_ring_ptr sq_ring_ptr, sqes_ptr sqes_ptr, 		sqe ptr[in, io_uring_sqe],   sqes_index int32)
+	// C:       syz_io_uring_submit(char* sq_ring_ptr,       io_uring_sqe* sqes_ptr,    io_uring_sqe* sqe,           uint32 sqes_index)
+
+	// It is not checked if the ring is full
+
+	// Cast to original
+	char* sq_ring_ptr = (char*)a0; // This will be exposed to offsets in bytes
+	char* sqes_ptr = (char*)a1;
+	char* sqe = (char*)a2;
+	uint32 sqes_index = (uint32)a3;
+
+	uint32 sq_ring_entries = *(uint32*)(sq_ring_ptr + SQ_RING_ENTRIES_OFFSET);
+	uint32 cq_ring_entries = *(uint32*)(sq_ring_ptr + CQ_RING_ENTRIES_OFFSET);
+
+	// Compute the sq_array offset
+	uint32 sq_array_off = SQ_ARRAY_OFFSET(sq_ring_entries, cq_ring_entries);
+
+	// Get the ptr to the destination for the sqe
+	if (sq_ring_entries)
+		sqes_index %= sq_ring_entries;
+	char* sqe_dest = sqes_ptr + sqes_index * SIZEOF_IO_URING_SQE;
+
+	// Write the sqe entry to its destination in sqes
+	memcpy(sqe_dest, sqe, SIZEOF_IO_URING_SQE);
+
+	// Write the index to the sqe array
+	uint32 sq_ring_mask = *(uint32*)(sq_ring_ptr + SQ_RING_MASK_OFFSET);
+	uint32* sq_tail_ptr = (uint32*)(sq_ring_ptr + SQ_TAIL_OFFSET);
+	uint32 sq_tail = *sq_tail_ptr & sq_ring_mask;
+	uint32 sq_tail_next = *sq_tail_ptr + 1;
+	uint32* sq_array = *(uint32**)(sq_ring_ptr + sq_array_off);
+	sq_array[sq_tail] = sqes_index;
+
+	// Advance the tail. Tail is a free-flowing integer and relies on natural wrapping.
+	// Ensure that the kernel will never see a tail update without the preceeding SQE
+	// stores being done.
+	__atomic_store_n(sq_tail_ptr, sq_tail_next, __ATOMIC_RELEASE);
+
+	// Now the application is free to call io_uring_enter() to submit the sqe
+	return 0;
+}
+
+#endif
+
+#if SYZ_EXECUTOR || __NR_syz_io_uring_put_ring_metadata
+
+// Used for syz_io_uring_put_ring_metadata(). Currently, all values are uint32,
+// therefore, usage of template type parameter type_t does not contribute much.
+#define put_val_into_ring(ring_ptr, offset, type_t, val) (*(type_t*)(ring_ptr + offset) = (type_t)val)
+
+static long syz_io_uring_put_ring_metadata(volatile long a0, volatile long a1, volatile long a2, volatile long a3)
+{
+	// syzlang: syz_io_uring_put_ring_metadata(string ring_type, ring_ptr ring_ptr, string field_name, T field_val)
+	// C:       syz_io_uring_put_ring_metadata(const char* ring_type, char* ring_ptr, const char* field_name, T field_val)
+
+	// Cast to original
+	const char* ring_type = (char*)a0;
+	const char* ring_ptr = (char*)a1;
+	const char* field_name = (char*)a2;
+
+	// Just map the ring type and field name to the offset and put the value
+	if (!strcmp(ring_type, "sq")) {
+		// sq ring
+		if (!strcmp(field_name, "head")) {
+			put_val_into_ring(ring_ptr, SQ_HEAD_OFFSET, uint32, a3);
```

Comment:
All offsets are constant, so we could as well move all of this into the descriptions. Namely have a function that is given based pointer, offset and value to write.
Moreover, sq ring and cq ring are actually the same :)
https://elixir.bootlin.com/linux/v5.8-rc4/source/fs/io_uring.c#L7548
If we hardcode all offsets, then hardcoding this assumption as well won't make things worse.
So we could pass, say, sq_ring_ptr and these offsets 0, 64, 256, 264, .... and a value.
The downside is that we will duplicate these offsets across executor and descriptions.
But the benefit is that maybe we could reuse this call for other similar subsystems (e.g. VHOST_SET_VRING_ADDR or XDP_PGOFF_RX_RING).

I hate making such design decisions...

Reply:
Okay, this looks like memcpy that accepts offsets (and felt like there should've been one, or is there?). Thus, I did this a bit more generic and implemented `syz_memcpy_off() (check [here](https://github.com/google/syzkaller/pull/1926/commits/95ef482c2fc9666dc755b25494fee45190a89ae2))`. Does it look reasonable to you?

Reply:
Looks reasonable. Let's go with this version.